src/routes/coingeckoProxy/index.ts

```diff
@@ -0,0 +1,93 @@
+import fastifyCaching from "@fastify/caching";
+import httpProxy from "@fastify/http-proxy";
+import { FastifyPluginAsync } from "fastify";
+import { ReadableStream } from "stream/web";
+
+const CACHE_TTL = 5 * 60 * 1000; // 5 min in ms
+
+type CachedData = {
+  item: {
+    contents: string;
+    headers: Headers;
+  };
+};
+
+function isCachedResponse(cached: any): cached is CachedData {
+  return (
+    cached &&
+    typeof cached === "object" &&
+    "item" in cached &&
+    typeof cached.item === "object" &&
+    "contents" in cached.item
+  );
+}
+
+const coingeckoProxy: FastifyPluginAsync = async (
+  fastify,
+  opts
+): Promise<void> => {
+  // Set up caching
+  fastify.register(fastifyCaching, { expiresIn: 300, serverExpiresIn: 300 });
+
+  // Intercepts request and uses cache if found
+  fastify.addHook("onRequest", async function (req, reply) {
+    fastify.cache.get(req.url.toLowerCase(), (err, result) => {
+      if (err) {
+        fastify.log.warn(`Error fetching cache for '${req.url}'`, err);
+        return err;
+      }
+      if (isCachedResponse(result)) {
+        fastify.log.info(`onRequest: cache hit for ${req.url}`);
+        reply
+          .headers({ "cache-hit": true, ...result.item.headers })
+          .send(result.item.contents);
+      }
+    });
+  });
+
+  // Intercepts response to store result on cache
+  fastify.addHook("onSend", async function (req, reply, payload) {
+    if (reply.getHeader("cache-hit")) {
+      // Header set when there's a cache hit, remove it
+      reply.removeHeader("cache-hit");
+    } else if (reply.statusCode >= 200 && reply.statusCode < 300) {
+      fastify.log.info(`onSend: caching response for ${req.url}`);
+
+      // No cache hit, consume the payload stream to be able to cache it
+      let contents = "";
+      for await (const chunk of payload as ReadableStream) {
+        contents += chunk.toString(); // Process each chunk of data
+      }
+
+      // Cache contents and headers
+      fastify.cache.set(
+        req.url.toLowerCase(),
+        { contents, headers: reply.getHeaders() },
+        CACHE_TTL,
+        (error) => {
+          if (error)
+            fastify.log.error(`Failed to cache result for ${req.url}`, error);
+        }
+      );
+
+      // Return consumed payload
+      // This is important!!! Without this, the response will be empty
+      return contents;
+    }
+  });
+
+  fastify.register(httpProxy, {
+    upstream: "https://api.coingecko.com",
+    rewritePrefix: "/api/v3/simple/",
```

Comment:
This means:
- Take the original url and replace it with `https://api.coingecko.com` + `/api/v3/simple/` + `<request params>`

Example:
From `http://localhost:3000/api/serverless/coingeckoProxy/` + `token_price/ethereum?contract_addresses=0x035dF12E0F3ac6671126525f1015E47D79dFEDDF&vs_currencies=usd`
To `https://api.coingecko.com/api/v3/simple/` + `token_price/ethereum?contract_addresses=0x035dF12E0F3ac6671126525f1015E47D79dFEDDF&vs_currencies=usd`